### PR TITLE
Refine CUDA Stream priority

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143849
* #143799
* #141123
* #141119
* #142347

# Motivation
As mentioned in https://github.com/pytorch/pytorch/pull/141119#discussion_r1897480515, we properly handle the priority value if it is outside of the priority range.

# Additional Context
If the value falls outside of the allowed priority range, it will automatically be mapped to the nearest valid priority(either lowest or highest).

### CODE REVIEW

File: c10/cuda/CUDAStream.cpp [hunk 1]
```diff
@@ -318,20 +318,14 @@ CUDAStream getStreamFromPool(const int priority, DeviceIndex device_index) {
...
 #if !defined(USE_ROCM)
   // See Note [HIP Lazy Streams]
   // CUDA-only: Initializes the stream pools (once)
   c10::call_once(
       device_flags[device_index], initDeviceStreamState, device_index);
 #endif
-  auto pri_idx = -priority;
-  pri_idx =
-      std::min(pri_idx, max_stream_priorities - 1); // pri_idx is zero-based
+  auto pri_idx = std::max(std::min(-priority, max_stream_priorities - 1), 0);
```
Comment (Skylion007) [inline c10/cuda/CUDAStream.cpp:328]:
```suggestion
  auto pri_idx = std::clamp(-priority, 0, max_stream_priorities - 1);
```

**Review by Skylion007 — COMMENTED** [review]
Use `std::clamp` as it's available in C++17, is more readable, and may be faster.

File: c10/cuda/CUDAStream.cpp (hunk 1 above)
Comment (guangyey) [reply c10/cuda/CUDAStream.cpp]: Make sense. Updated. Thanks.

File: torch/cuda/streams.py [hunk 2]
```diff
@@ -22,9 +22,10 @@ class Stream(torch._C._CudaStreamBase):
         device(torch.device or int, optional): a device on which to allocate
             the stream. If :attr:`device` is ``None`` (default) or a negative
             integer, this will use the current device.
-        priority(int, optional): priority of the stream, should be 0 or
-            negative, where negative numbers indicate higher priority. By default,
-            streams have priority 0.
+        priority(int, optional): priority of the stream, which can be positive, 0, or negative.
+            A lower number indicates a higher priority. By default, the priority is set to 0.
+            If the value falls outside of the allowed priority range, it will automatically be
+            mapped to the nearest valid priority(either lowest or highest).
```
Comment (albanD) [inline torch/cuda/streams.py:28]: Same nit as the xpu version.

**Review by albanD — APPROVED** [review]
SGTM

File: torch/cuda/streams.py (hunk 2 above)
Comment (guangyey) [reply torch/cuda/streams.py]: Thanks, updated.

**Review by EikanWang — APPROVED** [review]

**Pull request closed without merging**